include/rmm/mr/device/logging_resource_adaptor.hpp

```diff
@@ -0,0 +1,194 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "device_memory_resource.hpp"
+
+#include <rmm/detail/error.hpp>
+
+#include <sstream>
+#include <spdlog/sinks/basic_file_sink.h>
+#include <spdlog/spdlog.h>
+
+namespace rmm {
+namespace mr {
+/**
+ * @brief Resource that uses `Upstream` to allocate memory and logs information
+ * about the requested allocation/deallocations.
+ *
+ * An instance of this resource can be constructured with an existing, upstream
+ * resource in order to satisfy allocation requests and log
+ * allocation/deallocation activity.
+ *
+ * @tparam Upstream Type of the upstream resource used for
+ * allocation/deallocation.
+ */
+template <typename Upstream>
+class logging_resource_adaptor final : public device_memory_resource {
+ public:
+  /**
+   * @brief Construct a new logging resource adaptor using `upstream` to satisfy
+   * allocation requests and logging information about each allocation/free to
+   * the file specified by `filename`.
+   *
+   * The logfile will be written using CSV formatting.
+   *
+   * If the directories specified in `filename` do not exist, they will be
+   * created.
+   *
+   * Truncates `filename` if it already exists.
+   *
+   * Creating multiple `logging_resource_adaptor`s with the same `filename` will
+   * result in undefined behavior.
+   *
+   * @throws `rmm::logic_error` if `upstream == nullptr`
+   * @throws `spdlog::spdlog_ex` if opening `filename` failed
+   *
+   * @param upstream The resource used for allocating/deallocating device memory
+   * @param filename Name of file to write log info. If not specified, retrieves
+   * the file name from the environment variable "RMM_LOG_FILE".
+   */
+  logging_resource_adaptor(Upstream* upstream, std::string const& filename =
+                                                   std::getenv("RMM_LOG_FILE"))
+      : upstream_{upstream},
+        logger_{std::make_shared<spdlog::logger>(
+            "RMM", std::make_shared<spdlog::sinks::basic_file_sink_mt>(
+                       filename, true /*truncate file*/))} {
+    RMM_EXPECTS(nullptr != upstream,
+                "Unexpected null upstream resource pointer.");
+
+    auto const csv_header{"Time,Action,Pointer,Size,Stream"};
+    logger_->set_pattern("%v");
+    logger_->info(csv_header);
+    logger_->set_pattern("%H:%M:%S:%f,%v");
+  }
+
+  bool supports_streams() const noexcept override {
+    return upstream_->supports_streams();
+  }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream
+   * resource and logs the allocation.
+   *
+   * If the upstream allocation is successful logs the
```

Comment:
Hmmm, would it be too much overhead to catch allocation failure exceptions and log them, then throw the same exception?

Reply:
I think we can explore that as a different level of logging. I think it's valuable to have a level of logging that only logs successful allocations. Otherwise I'd have to add the ability to detect exceptions in the log in the replay benchmark.